src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/AnalyzerWorkQueue.cs

```diff
@@ -12,98 +12,118 @@ namespace OmniSharp.Roslyn.CSharp.Workers.Diagnostics
 {
     public class AnalyzerWorkQueue
```

Comment:
Is this something that a [`BlockingCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1) might make a little more straight forward?  

I've used this in the past to deal with spinning and waiting for things https://github.com/OmniSharp/csharp-language-server-protocol/blob/9fa60ffdaf7762ef29f8badea6182f272d0e5979/src/JsonRpc/ProcessScheduler.cs#L73

Reply:
The big win is you can avoid using locks because a Blocking collection will handle it for you.

Reply:
Tried to implement with BlockinCollection but got to dead end because theres throttling and timeout mechamism (it's assertion after this PR but anyway) that still require locking. Another problematic part was that in large workspaces there may be tens of thousands of files and on those situation cases like promote mechanism needs to loop all files through to find is there match or not in collection, which is reason why hashset is used originally.

RX could be good fit to this since theres time based throttling and producer/consumer patterns combined 🤔 But i could say lets leave it for further refactoring since this PR already simplifies queue mechanism?

Reply:
Yeah, I'm fine with bringing in RX later to help with that.  I was just wondering it might help with the scenario or not. 👍